### PR TITLE
Test Case for Incr. Comp. Hash for enums #36674.

### LINKED ISSUE #36674: incr.comp.: Create Test Case for Incr. Comp. Hash for enums
This issue is part of a broader effort to implement comprehensive regression testing for incremental compilation. For the tracking issue go to #36350.

### Background
For incremental compilation we need to determine if a given HIR node has changed in between two versions of a program. This is implemented in the [calculate_svh](https://github.com/rust-lang/rust/tree/master/src/librustc_incremental/calculate_svh) module. We compute a hash value for each HIR node that corresponds to a node in the dependency graph and then compare those hash values. We call this hash value the *Incremental Compilation Hash* (ICH) of the HIR node. It is supposed to be sensitive to any change that might make a semantic difference to the thing being hashed.

### Testing Methodology
The auto-tests in the `src/test/incremental` directory all follow the same pattern:

* Each source file defines one test case
* The source file is compiled multiple times with incremental compilation turned on, each time with a different `--cfg` flag, allowing each version to differ via conditional compilation. Each of these versions we call a *revision*.
* During each revision, the test runner will make sure that some assertions about the dependency graph of the program hold.
* These assertions are specified in the source code of the test file as attributes attached to the items being tested (e.g. `#[rustc_clean]`/`#[rustc_dirty]`).

The ICH-specific tests use this framework by adhering to the following pattern:

* There are two versions of the definition under test, one marked with `#[cfg(cfail1)]` and the second marked with `#[cfg(not(cfail1))]`. As a consequence, the first revision will be compiled with the first version of the definition, and all other revisions will be compiled with the second version. The two versions are supposed to differ in exactly one aspect (e.g. the visibility of a field is different, or the return of a function has changed).
* The second definition has a `#[rustc_dirty(label="Hir", cfg="cfail2")]` attribute attached. This attribute makes the test runner check that a change of the `Hir` dependency node of the definition has been detected between revisions `cfail1` and `cfail2`. This will effectively test that a different ICH value has been computed for the two versions of the definition.
* The second definition also has a `#[rustc_clean(label="Hir", cfg="cfail3")]` attribute. This makes sure that the `Hir` dependency node (and thus the ICH) of the definition has **not** changed between revisions `cfail2` and `cfail3`. That's what we expect, because both revisions use the same version of the definition.
* For definitions that are exported from the crate, we also want to check the ICH of the corresponding metadata. This is tested using the `#[rustc_metadata_clean]`/`#[rustc_metadata_dirty]` attributes and works analogous to the `Hir` case: We add `#[rustc_metadata_dirty(cfg="cfail2")]` to the second definition to make sure that the ICH of the exported metadata is not the same for the different versions of the definition, and we add `#[rustc_metadata_dirty(cfg="cfail3")]` to make sure that the ICH is the same for the two identical versions of the definition.

Why are the revisions called "cfail"? That's because of reasons internal to how
the test-runner works. Prefixing a revision with "cfail" tells the test runner to treat the test like a "compile-file" test, that is: compile the test case but don't actually run it (which would be the case for an "rpass" revision). For the ICH tests we need to compile "rlibs", so that we can test metadata ICHs. As a consequence we cannot declare them "rpass". In an additional directive (`// must-compile-successfully`), we tell the test runner that we actually expect the file to not produce any errors.

Each test case must contain the following test-runner directives and crate attributes at the top:
```rust
// must-compile-successfully
// revisions: cfail1 cfail2 cfail3
// compile-flags: -Z query-dep-graph

#![feature(rustc_attrs)]  // <-- let's us use #[rustc_dirty], etc.
#![crate_type="rlib"]     // <-- makes sure that we export definitions
```

See [src/test/incremental/hashes/struct_defs.rs](https://github.com/rust-lang/rust/blob/master/src/test/incremental/hashes/struct_defs.rs) for a full example of such a ICH regression test.

## Enum Specific ICH Testing

Each of the following things should be tested with its own definition pair:

* Change enum visibility
* Change name of a c-style variant
* Change name of a tuple-style variant
* Change name of a struct-style variant
* Change the value of a c-style variant
* Add a c-style variant
* Remove a c-style variant
* Add a tuple-style variant
* Remove a tuple-style variant
* Add a struct-style variant
* Remove a struct-style variant
* Change the type of a field in a tuple-style variant
* Change the type of a field in a struct-style variant
* Change the name of a field in a struct-style variant
* ~~Change the visibility of a field in a tuple-style variant~~ N/A for enums
* ~~Change the visibility of a field in a struct-style variant~~ N/A for enums
* Change order of fields in a tuple-style variant
* Change order of fields in a struct-style variant
* Add a field to a tuple-style variant
* Add a field to a struct-style variant
* Add #[must_use] to the enum
* Add #[repr(C)] to the enum
* Change the name of a type parameter
* Add a type parameter
* Change the name of a lifetime parameter
* Add a lifetime parameter
* Add a lifetime bound to a lifetime parameter
* Add a lifetime bound to a type parameter
* Add a trait bound to a type parameter
* Add a lifetime bound to a lifetime parameter in where clause
* Add a lifetime bound to a type parameter in where clause
* Add a trait bound to a type parameter in where clause
* In an enum with two variants, swap usage of type parameters
* In an enum with two variants, swap usage of lifetime parameters

EDIT: Some more cases
* Change field type in tuple-style variant [indirectly](https://github.com/michaelwoerister/rust/blob/e82462523b5ac032e9a28e0e35a9d7fdc318d864/src/test/incremental/hashes/struct_defs.rs#L243-L258) by modifying a `use` statement
* Change field type in record-style variant [indirectly](https://github.com/michaelwoerister/rust/blob/e82462523b5ac032e9a28e0e35a9d7fdc318d864/src/test/incremental/hashes/struct_defs.rs#L243-L258) by modifying a `use` statement
* Change trait bound of type parameter [indirectly](https://github.com/michaelwoerister/rust/blob/e82462523b5ac032e9a28e0e35a9d7fdc318d864/src/test/incremental/hashes/struct_defs.rs#L280-L295) by modifying a `use` statement
* Change trait bound of type parameter in where clause [indirectly](https://github.com/michaelwoerister/rust/blob/e82462523b5ac032e9a28e0e35a9d7fdc318d864/src/test/incremental/hashes/struct_defs.rs#L280-L295) by modifying a `use` statement

### PR DESCRIPTION
**Pull request by eulerdisk**
Fixes #36674
Part of #36350

r? @michaelwoerister

### CODE REVIEW

File: src/test/incremental/hashes/enum_defs.rs [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+    Variant1 { a: u32, b: f32 },
+}
+
+#[cfg(not(cfail1))]
+#[rustc_dirty(label="Hir", cfg="cfail2")]
+#[rustc_clean(label="Hir", cfg="cfail3")]
+#[rustc_metadata_dirty(cfg="cfail2")]
+#[rustc_metadata_clean(cfg="cfail3")]
+enum EnumChangeFieldOrderStructStyleVariant {
+    Variant1 { b: u32, a: u32 },
```
Comment (michaelwoerister) [inline src/test/incremental/hashes/enum_defs.rs]: `b` should be `f32`

File: src/test/incremental/hashes/enum_defs.rs [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+// for struct definitions.
+
+// The general pattern followed here is: Change one thing between rev1 and rev2
+// and make sure that the hash has changed, then change nothing between rev2 and
+// rev3 and make sure that the hash has not changed.
+
+// We also test the ICH for struct definitions exported in metadata. Same as
+// above, we want to make sure that the change between rev1 and rev2 also
+// results in a change of the ICH for the struct's metadata, and that it stays
+// the same between rev2 and rev3.
```
Comment (michaelwoerister) [inline src/test/incremental/hashes/enum_defs.rs]: The comment paragraph above still says 'struct' instead of 'enum'.

**Review by michaelwoerister — CHANGES_REQUESTED** [review]
This is excellent, thank you so much, @eulerdisk! 
I've left two comments on things that should be fixed. After that this is good to merge :+1:

**Comment by michaelwoerister** [comment]
@bors r+

Thanks @eulerdisk! 
(FYI: I used GH's new editing feature to make another "struct" -> "enum" correction in a comment)

**Comment by eulerdisk** [comment]
@michaelwoerister Maybe there's another "struct" to change in line 21

**Comment by michaelwoerister** [comment]
Yes, you're right, I'll fix it.

**Comment by michaelwoerister** [comment]
:tada:

**Pull request merged**